core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1019 @@
+      // we decide to eject the subchannel.
+      OutlierDetectionSubchannel subchannel = new OutlierDetectionSubchannel(
+          delegate.createSubchannel(args));
+
+      // If the subchannel is associated with a single address that is also already in the map
+      // the subchannel will be added to the map and be included in outlier detection.
+      List<EquivalentAddressGroup> allAddresses = subchannel.getAllAddresses();
+      if (allAddresses.size() == 1 && trackerMap.containsKey(allAddresses.get(0))) {
+        AddressTracker eagInfo = trackerMap.get(allAddresses.get(0));
+        subchannel.setEquivalentAddressGroupInfo(eagInfo);
+        eagInfo.addSubchannel(subchannel);
+
+        // If this address has already been ejected, we need to immediately eject this Subchannel.
+        if (eagInfo.ejectionTimeNanos != null) {
+          subchannel.eject();
+        }
+      }
+
+      return subchannel;
+    }
+
+    @Override
+    public void updateBalancingState(ConnectivityState newState, SubchannelPicker newPicker) {
+      delegate.updateBalancingState(newState, new OutlierDetectionPicker(newPicker));
+    }
+  }
+
+  class OutlierDetectionSubchannel extends ForwardingSubchannel {
+
+    private final Subchannel delegate;
+    private AddressTracker eagInfo;
+    private boolean ejected;
+    private ConnectivityStateInfo lastSubchannelState;
+    private OutlierDetectionSubchannelStateListener subchannelStateListener;
+
+    OutlierDetectionSubchannel(Subchannel delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public void start(SubchannelStateListener listener) {
+      subchannelStateListener = new OutlierDetectionSubchannelStateListener(listener);
+      super.start(subchannelStateListener);
+    }
+
+    @Override
+    public Attributes getAttributes() {
+      return delegate.getAttributes().toBuilder().set(EAG_INFO_ATTR_KEY, eagInfo).build();
+    }
+
+    @Override
+    public void updateAddresses(List<EquivalentAddressGroup> addresses) {
+      // Outlier detection only supports subchannels with a single address, but the list of
+      // addresses associated with a subchannel can change at any time, so we need to react to
+      // changes in the address list plurality.
+
+      // No change in address plurality, we replace the single one with a new one.
+      if (getAllAddresses().size() == 1 && addresses.size() == 1) {
+        // Remove the current subchannel from the old address it is associated with in the map.
+        if (trackerMap.containsKey(getAddresses())) {
+          trackerMap.get(getAddresses()).removeSubchannel(this);
+        }
+
+        // If the map has an entry for the new address, we associate this subchannel with it.
+        EquivalentAddressGroup newAddress = Iterables.getOnlyElement(addresses);
+        if (trackerMap.containsKey(newAddress)) {
+          AddressTracker tracker = trackerMap.get(newAddress);
+          tracker.addSubchannel(this);
```

Comment:
Don't we also need to remove the entry for this subchannel is the older AddressTracker?

Reply:
There is a removeSubchannel() call a few lines above for this.